proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+last case, however, currently, the `a` would refer to the type variable, but if
+Haskell had a single namespace it would refer to the term-level variable. Thus the
+warning is triggered.
+
+#### `-Wpuns`, example #4
+
+```haskell
+f :: [] a -- warning
+g :: [a]  -- warning
+g = []    -- warning
+h = [a]   -- warning
+x = [a,b] -- no warning
+```
+
+Since `Data.BuiltInTypes` is imported by default, all of the cases except the
+very last one will emit `-Wpuns` warning because in all of them it is not clear
+whether the data constructor or a type constructor is being referred to, except
+in the very last case.
+
+#### `-Wpuns`, example #5
+
+```haskell
+f :: ()      -- warning
+f = ()       -- warning
+g :: (a,b)   -- warning
+g = (c,d)    -- warning
+h :: (,) a b -- warning
+h = (,) c d  -- warning
+```
+
+Tuples in this case are very much the same as lists except they will emit a
+warning in all cases.
+
+Note that for both lists and tuples if `-XNoImplicitBuiltInTypes` is enabled,
+the type constructors will not be in scope anymore and no warnings will be
+emitted.
+
+#### `-Wpun-bindings`, example #1
+
+(This example uses [GHC Proposal #155](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0155-type-lambda.rst)).
+
+```haskell
+id :: t -> t
+id @a a = a
+```
+
+Here, when term-level `a` is bound it would conflict with the type level `a` if
+Haskell had a single namespace, thus triggering the warning. This behavior is
+similar to conflicting definition error for `f b b = ...`:
+
+```haskell
+Test.hs:1:3: error:
+    • Conflicting definitions for 'b'
+      Bound at: Test.hs:1:3
+                Test.hs:1:5
+    • In an equation for 'f'
+```
+
+On the contrary, the code below is fine, similarly to `-Wpuns` example #2, the `a` is shadowed instead:
+
+```haskell
+f :: t -> ()
+f @a = \a -> ()
+```
+
+#### `-Wpun-bindings`, example #2
+
+```haskell
+data T = T
+```
+
+If Haskell had a single unified namespace, type constructor `T` and data
+constructor `T` would conflict, thus this should trigger the warning.
+
+#### `-Wpun-bindings`, example #3
+
+```haskell
+data T = MkT
+data B = T | F
+```
+
+Even though type constructor `T` and data constructor `T` are defined in
+different declarations, they would still cause a conflict, same as example #2.
+
+#### `-Wpun-bindings`, example #4
+
+```haskell
+data J = Bool
+```
+
+This should not cause the warning because `Bool` defined here would not
+conflict with `Bool` imported from `Prelude`, this declaration is not rejected
+by GHC:
+
+```haskell
+import Prelude (Bool)
+data Bool -- no conflict
+```
+
+## Effect and Interactions
+
+* Users will be able to make sure their code is pun-free as well as have all
+  the necessary tools to interact with the code that does use punning (for
+  example, with libraries)
+
+* With the deprecation of `'` in `-XDataKinds` the syntax will be exclusively
+  used by `-XTemplateHaskell`.
+
+## Costs and Drawbacks
```

Comment:
An additional cost and drawback is that this proposal will cause ecosystem-wide breakage with the addition of `-Wpuns` and `-Wpun-bindings` in `-Wall`. I would expect well more than half of Haskell packages on Hackage would need updates.... quite possibly almost all packages. I predict it will be so bad that any reader would be able to tell, at a glance, whether a Haskell file was written before or after this change.

All of this is to say: I don't think we should ever put `-Wpuns` and `-Wpun-bindings` into `-Wall`. Or at least not for 20 years or so.